src/main/java/org/truffleruby/core/proc/ProcNodes.java

```diff
@@ -256,19 +258,32 @@ protected Object sourceLocation(RubyProc proc) {
 
         @Specialization
         protected RubyProc createSameArityProc(RubyProc userProc, RubyProc block) {
-            final RubyProc composedProc = new RubyProc(
-                    coreLibrary().procClass,
-                    getLanguage().procShape,
-                    block.type,
+            final RubyProc composedProc = block.withSharedMethodInfo(
                     userProc.sharedMethodInfo,
-                    block.callTargets,
-                    block.callTarget,
-                    block.declarationFrame,
-                    block.declarationVariables,
-                    block.method,
-                    block.block,
-                    block.frameOnStackMarker,
-                    block.declarationContext);
+                    coreLibrary().procClass,
+                    getLanguage().procShape);
+            AllocationTracing.trace(composedProc, this);
+            return composedProc;
+        }
+    }
+
+    @Primitive(name = "proc_specify_arity", lowerFixnum = 1)
+    public abstract static class ProcSpecifyArityNode extends PrimitiveArrayArgumentsNode {
+        @Specialization
+        protected RubyProc specifyArityProc(RubyProc block, int argc) {
+            Arity oldArity = block.sharedMethodInfo.getArity();
+            final Arity newArity;
+            if (argc <= -1) {
+                newArity = new Arity(0, 0, true);
```

Comment:
The first argument here should be `-(argc + 1)`. It doesn't matter for the `rb_define_method` case, but this might be used in other places, and there e.g. `-3` should mean 2 required arguments + any nb of arguments (rest).